main.go

```diff
@@ -178,24 +179,158 @@ func isPregnant(kittyID uint64) (bool, error) {
 }
 
-func (c *KittyContract) CanBreedWith(ctx contractapi.TransactionContextInterface, sire uint64, matron uint64) (bool, error) {
+func (c *KittyContract) CanBreedWith(ctx contractapi.TransactionContextInterface, sireID, matronID uint64) (bool, error) {
+	if int(matronID) >= len(kitties) || matronID == 0 {
+		return false, fmt.Errorf("No matron with this ID %d available.", matronID)
+	}
+
+	if int(sireID) >= len(kitties) || sireID == 0 {
+		return false, fmt.Errorf("No sire with this ID %d available.", sireID)
+	}
+
+	ok, err := isValidMatingPair(matronID, sireID)
+	if err != nil || !ok {
+		return false, err
+	}
+
+	ok, err = isSiringPermitted(matronID, sireID)
+	if err != nil || !ok {
+		return false, err
+	}
+
 	return true, nil
 }
 
-func breedWith(sire uint64, matron uint64) error {
+func breedWith(sireID, matronID uint64) error {
+	if int(matronID) >= len(kitties) {
+		return fmt.Errorf("No kitty with this ID %d available.", matronID)
+	}
+
+	if int(sireID) >= len(kitties) {
+		return fmt.Errorf("No kitty with this ID %d available.", sireID)
+	}
+
+	matron := kitties[matronID]
+
+	matron.SiringWithID = sireID
+	triggerCooldown(matronID)
+	triggerCooldown(sireID)
+
+	sireAllowedToAddress[sireID] = ""
+	sireAllowedToAddress[matronID] = ""
+
+	owner := kittyIndexToOwner[matronID]
+
+	g_event["Pregnant"] = map[string]interface{}{"owner": owner, "matronID": matronID, "sireID": sireID, "matronCooldown": matron.CooldownEnd}
+
 	return nil
 }
 
-func (c *KittyContract) BreedWithAuto(ctx contractapi.TransactionContextInterface, sireID uint64, matronID uint64) error {
-	return nil
+func (c *KittyContract) BreedWithAuto(ctx contractapi.TransactionContextInterface, sireID, matronID uint64) error {
+	if int(matronID) >= len(kitties) {
+		return fmt.Errorf("No kitty with this ID %d available.", matronID)
+	}
+
+	if int(sireID) >= len(kitties) {
+		return fmt.Errorf("No kitty with this ID %d available.", sireID)
+	}
+
+	clientID, err := ctx.GetClientIdentity().GetID()
+	if err != nil {
+		return err
+	}
+
+	matronOwner := kittyIndexToOwner[matronID]
+
+	if matronOwner != clientID {
+		return fmt.Errorf("Caller must be the owner of the matron kitty.")
+	}
+
+	if ok, err := isReadyToBreed(matronID); err != nil || !ok {
+		return fmt.Errorf("Provided marton with id %d is not ready to breed.", matronID)
+	}
+
+	if ok, err := isReadyToBreed(sireID); err != nil || !ok {
+		return fmt.Errorf("Provided sire with id %d is not ready to breed.", sireID)
+	}
+
+	if ok, err := isValidMatingPair(sireID, matronID); err != nil || !ok {
+		return fmt.Errorf("Matron with id %d and sire with id %d are no valid mating pair. Shame on you! Don't try to breed these cats.", matronID, sireID)
+	}
+
+	return breedWith(sireID, matronID)
 }
 
-func (c *KittyContract) GiveBirth(ctx contractapi.TransactionContextInterface, matronID uint64) error {
-	return nil
+func mixGenes(matronGenes, sireGenes uint64) uint64 {
+	return matronGenes ^ sireGenes // ^ will implement xor
```

Comment:
xor is a very nice idea for the mixing! :D :+1:

Reply:
:)